waku/waku_api/message_cache.nim

```diff
@@ -15,68 +15,292 @@ import
+  self.contentTopics.del(contentIdx)
+  
+  var msgIndexes = newSeq[int](0)
+
+  var j = self.contentIndex.high
+  while j > -1:
+    let (cId, mId) = self.contentIndex[j]
+
+    if cId == contentIdx:
+      # remove indexes for this topic
+      self.contentIndex.del(j)
+      msgIndexes.add(mId)
+    elif cId == lastIndex:
+      # swap the indexes because contentTopics.del() is a swap
+      self.contentIndex[j] = (contentIdx, mId)
+    
+    dec(j)
+
+  # check if messages on this content topic are indexed by any pubsub topic, if not remove them.
+  for mId in msgIndexes:
+    if not self.pubsubIndex.anyIt(it[1] == mId):
+      self.removeMessage(mId)
+
+proc reset*(self: MessageCache) =
+  self.writeIndex = 0
+  self.messages.setLen(0)
+  self.pubsubTopics.setLen(0)
+  self.contentTopics.setLen(0)
+  self.pubsubIndex.setLen(0)
+  self.contentIndex.setLen(0)
+
+proc addMessage*(
+  self: MessageCache,
+  pubsubTopic: PubsubTopic,
+  msg: WakuMessage
+  ) =
+  ## Idempotent message addition.
+  
+  for i, message in self.messages:
+    if message == msg:
+      return
+
+  var pubsubIdxOp = self.pubsubSearch(pubsubTopic)
+  var contentIdxOp = self.contentSearch(msg.contentTopic)
+
+  if pubsubIdxOp.isNone() and contentIdxOp.isNone():
     return
 
-  # Make a copy of msgs for this topic to modify
-  var messages = t.table.getOrDefault(topic, @[])
+  let pubsubIdx =
+    if pubsubIdxOp.isNone():
+      self.pubsubTopics.add(pubsubTopic)
+      self.pubsubTopics.high
+    else:
+      pubsubIdxOp.get()
+
+  let contentIdx =
+    if contentIdxOp.isNone():
+      self.contentTopics.add(msg.contentTopic)
+      self.contentTopics.high
+    else:
+      contentIdxOp.get()
 
-  if messages.len >= t.capacity.int:
-    trace "Topic cache capacity reached", topic=topic
-    # Message cache on this topic exceeds maximum. Delete oldest.
-    # TODO: this may become a bottle neck if called as the norm rather than
-    #  exception when adding messages. Performance profile needed.
-    messages.delete(0,0)
+  # add the message, make space if needed
+  if self.messages.len >= self.capacity:
+    # replace the oldest message
+    self.messages[self.writeIndex] = msg
 
-  messages.add(msg)
+    # replace the indexes of the replaced message
+    for (pId, mId) in self.pubsubIndex.mitems:
+      if mId == self.writeIndex:
+        pId = pubsubIdx
 
-  # Replace indexed entry with copy
-  t.table[topic] = messages
+    for (cId, mId) in self.contentIndex.mitems:
+      if mId == self.writeIndex:
+        cId = contentIdx
 
-proc clearMessages*[K](t: MessageCache[K], topic: K) =
-  if not t.isSubscribed(topic):
+    # update the write index
+    inc(self.writeIndex)
+    if self.writeIndex >= self.capacity:
+      self.writeIndex = 0
+    
     return
-  t.table[topic] = @[]
+ 
+  let msgIdx = self.messages.len
+  self.messages.add(msg)
+
+  self.pubsubIndex.add((pubsubIdx, msgIdx))
+  self.contentIndex.add((contentIdx, msgIdx))
```

Comment:
Do we need to separate it in two cases?
I think that the code inside the `self.messages.len >= self.capacity` block would work in both cases

Reply:
Is it possible to `self.messages[self.messages.len] = msg` ?

Reply:
Oh sorry, just saw your response. My bad then :)